service/server.py

```diff
@@ -32,59 +39,111 @@
+
+    query = PPI_QUERY_TMPL.format('\n'.join(kv_lines))
     ppi_url = ppi_api
     resp = requests.post(ppi_url, data={'output': 'json', 'query': query})
 
     sale_list = resp.json()['results']['bindings']
 
     if len(sale_list) == 0:
-        return jsonify({'message': 'No record found.'})
-    elif len(sale_list) > 1:
-        return jsonify({'message': 'More then one record found.'})
+        return {}
+
+    latest_sale = {k: v['value'] for k, v in sale_list[0].items()}
+    return latest_sale
 
-    latest_sale = sale_list[0]
+
+def get_property_address(query_dict):
+    results = (AddressBase.query.
+        filter_by(postcode=query_dict['postcode']).
+        filter_by(thoroughfareName=query_dict['street']).
+        filter(or_(
+            AddressBase.buildingNumber == query_dict['paon'],
+            AddressBase.buildingName == query_dict['paon'])))
+    if 'saon' in query_dict:
+        results = results.filter_by(subBuildingName=query_dict['saon'])
+
+    return results.all()
+
+
+def create_json(address_rec, latest_sale):
+    # try and get buildingNumber, otherwise buildingName
+    paon = address_rec.buildingNumber or address_rec.buildingName.rstrip()
 
     result = {
-        'saon': 'saon goes here',
-        'paon': 'paon goes here',
-        'street': 'street goes here',
-        'town': 'town goes here',
-        'county': 'county goes here',
-        'postcode': 'postcode goes here',
-        'amount': latest_sale['amount']['value'],
-        'date': latest_sale['date']['value'],
+        'saon': address_rec.subBuildingName.rstrip(),
+        'paon': paon,
+        'street': address_rec.thoroughfareName.rstrip(),
+        'town': address_rec.postTown.rstrip(),
+        'county': address_rec.dependentLocality.rstrip(),
+        'postcode': address_rec.postcode.rstrip(),
+        'amount': latest_sale.get('amount', ''),
+        'date': latest_sale.get('date', ''),
         'property_type':
-            get_property_type(latest_sale['property_type']['value']),
-        'coordinates' : {'latitude': 99, 'longitude': 99},
+            get_property_type(latest_sale.get('property_type', '')),
+        'coordinates' : {
+            'latitude': address_rec.positionY,
+            'longitude': address_rec.positionX,
+        },
     }
+    return result
+
+
+@app.route('/properties/<postcode>/<street_paon_saon>', methods=['GET'])
+def get_property(postcode, street_paon_saon):
+    # TODO: remove this
+    if postcode == 'N1BLT' and street_paon_saon == 'imaginary-street':
+        abort(404)
+
+    field_vals = [postcode] + street_paon_saon.split('_')
+    check_field_vals(field_vals)
+    query_dict = get_query_dict(field_vals)
+
+    address_recs = get_property_address(query_dict)
+    nof_results = len(address_recs)
+    if nof_results == 0:
+        abort(404)
+    elif nof_results > 1:
+        raise NotImplementedError('More than one record found')
```

Comment:
What HTML status code should we have here? Is it even possible to have multiple records in the architecture @AndrewGorton has in mind?